include/functional/detail/not_tuple.hpp

```diff
@@ -8,47 +8,29 @@
 
 namespace fn::detail {
 
-template <typename... Args> struct not_tuple;
+template <std::size_t I, typename T> struct _element {
+  T v;
 
-template <typename Arg0> struct not_tuple<Arg0> final {
-  constexpr static unsigned size = 1;
-  Arg0 v0;
-
-  static_assert(!std::is_rvalue_reference_v<Arg0>);
+  static_assert(not std::is_rvalue_reference_v<T>);
 };
 
-template <typename Arg0, typename Arg1> struct not_tuple<Arg0, Arg1> final {
-  constexpr static unsigned size = 2;
-  Arg0 v0;
-  Arg1 v1;
-
-  static_assert(!std::is_rvalue_reference_v<Arg0>
-                && !std::is_rvalue_reference_v<Arg1>);
+template <typename, typename...> struct not_tuple_base;
+
+template <std::size_t... Is, typename... Ts>
+struct not_tuple_base<std::index_sequence<Is...>, Ts...> : _element<Is, Ts>... {
+  template <typename Self, typename Fn, typename... Args>
+  static auto invoke(Self &&self, Fn &&fn, Args &&...args) noexcept
+      -> decltype(auto)
+    requires std::invocable<
+        decltype(fn), decltype(args)...,
+        decltype(apply_const_t<Self, Ts>(self._element<Is, Ts>::v))...>
+  {
+    return std::forward<decltype(fn)>(fn)(
+        std::forward<decltype(args)>(args)...,
+        apply_const_t<Self, Ts>(self._element<Is, Ts>::v)...);
```

Comment:
also, why is this apply_const and not also reference? How is this going to be efficient without moves?

Reply:
good point. I fixed this and added test for immovable data.

Reply:
FWIW test https://github.com/Bronek/functional/pull/13/files#diff-a08665e0c9fe59f6c071bd1788645293b019b7e849bf6f553a682a63517feba4R211-R237